src/inmanta/server/agentmanager.py

```diff
@@ -147,14 +147,98 @@ def get_depended_by(self) -> List[str]:
     def expire(self, session: protocol.Session, timeout: float) -> None:
         self.add_background_task(self._expire_session(session, datetime.now()))
 
-    def seen(self, session: protocol.Session, endpoint_names: List[str]) -> None:
-        if set(session.endpoint_names) != set(endpoint_names):
-            LOGGER.warning(
-                "Agent endpoint set changed, this should not occur, update ignored (was %s is %s)"
-                % (set(session.endpoint_names), set(endpoint_names))
+    def seen(self, session: protocol.Session) -> None:
+        self.add_background_task(self._seen_session(session))
+
+    async def _seen_session(self, session: protocol.Session) -> None:
+        endpoints_with_new_primary: List[Tuple[str, Optional[protocol.Session]]] = []
+        async with self.session_lock:
+            endpoints_in_agent_manager = set(
```

Comment:
I would keep a proper data structure to cache this, and do the lookup O(1), it seems a bit excessive to go over every session we have every time we get a heartbeat in.